### PR TITLE
HPCC-14586 Support HINT on COUNT() and DATASET([])

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@jamienoss please review

File: ecl/hql/hqlutil.cpp [hunk 1]
```diff
@@ -4608,7 +4613,17 @@ extern HQL_API IHqlExpression * convertScalarAggregateToDataset(IHqlExpression *
...
 
-    IHqlExpression * project = createDataset(no_newaggregate, dataset, createComma(aggregateRecord, transform, LINK(keyedAttr), LINK(prefetchAttr)));
+    HqlExprArray args;
+    args.append(*dataset);
+    args.append(*aggregateRecord);
+    args.append(*transform);
+    ForEachChild(i2, expr)
+    {
+        IHqlExpression * cur = expr->queryChild(i2);
+        if (cur->isAttribute())
```
Comment (jamienoss) [inline ecl/hql/hqlutil.cpp]: Is it ok/intended to include the keyed attribute even though this is already/also included in valueArgs? Not sure if the only point to this loop is to ensure all attributes are preserved where they were ignored from valueArgs or...?

File: ecl/regress/count7.ecl [hunk 2]
```diff
@@ -39,5 +39,5 @@ output(count(namesTable5) = 5);
 namesTable6 := dataset('x6',namesRecord,FLAT);
 output(count(namesTable6) != 5);
 namesTable7 := dataset('x7',namesRecord,FLAT);
-output(count(namesTable7) != 5);
-output(count(namesTable7) > 0);
+output(count(namesTable7, HINT(goQuitefast(true))) != 5);
+output(count(namesTable7, HINT(goReallyRellyFast(true))) > 0);
```
Comment (jamienoss) [inline ecl/regress/count7.ecl]: Missing an 'a'.

File: ecl/regress/readahead2.ecl [hunk 3]
```diff
@@ -506,7 +506,7 @@ outrec it(in L) := TRANSFORM
   self := L;
 END;
 
-OUTPUT(SUM(in, TS_WordIndex(KEYED(word = in.word))[1].doc, prefetch(5, parallel))) : independent;
+OUTPUT(SUM(in, TS_WordIndex(KEYED(word = in.word))[1].doc, prefetch(5, parallel), hint(thisIsAHint('Help!')))) : independent;
```
Comment (jamienoss) [inline ecl/regress/readahead2.ecl]: 'hint' would be better in upper case.

**Comment by jamienoss** [comment]
@ghalliday back to you.

File: ecl/hql/hqlutil.cpp (hunk 1 above)
Comment (ghalliday) [inline ecl/hql/hqlutil.cpp]: Yes, it is desired.  Ideally keyed would not be added to the valueArgs, but that would require more substantial changes elsewhere.

**Comment by ghalliday** [comment]
@richardkchapman please review

**Comment by ghalliday** [comment]
and merge if happy.

**Pull request merged**